File: mobile/src/screens/profile/EditProfileScreen.js
import { useState } from 'react';
import { View, Text, Pressable, ScrollView, StyleSheet } from 'react-native';
import * as ImagePicker from 'expo-image-picker';
import Screen from '../../components/common/Screen';
import Input from '../../components/common/Input';
import Button from '../../components/common/Button';
import { useAuthStore } from '../../store/authStore';
import { updateMe } from '../../api/contacts.api';
import { uploadMedia } from '../../api/chat.api';
import { colors, radius } from '../../config/theme';

function linksToText(links) {
  return (links || []).map((l) => l.url).join('\n');
}
function textToLinks(text) {
  return text
    .split('\n')
    .map((line) => line.trim())
    .filter(Boolean)
    .map((url) => ({ label: url.replace(/^https?:\/\//, '').split('/')[0], url }));
}

const VISIBILITY_OPTIONS = [
  { id: 'public', label: 'Public' },
  { id: 'friends', label: 'Friends' },
  { id: 'only_me', label: 'Only me' },
];

export default function EditProfileScreen({ navigation }) {
  const user = useAuthStore((s) => s.user);
  const updateUser = useAuthStore((s) => s.updateUser);

  const [form, setForm] = useState({
    name: user?.name || '',
    bio: user?.bio || '',
    work: user?.work || '',
    education: user?.education || '',
    location: user?.location || '',
    links: linksToText(user?.links),
    profileVisibility: user?.profileVisibility || 'public',
  });
  const [saving, setSaving] = useState(false);
  const [uploadingAvatar, setUploadingAvatar] = useState(false);
  const [uploadingCover, setUploadingCover] = useState(false);

  const set = (key) => (value) => setForm((f) => ({ ...f, [key]: value }));

  const pickAndUpload = async (field, setBusy) => {
    const permission = await ImagePicker.requestMediaLibraryPermissionsAsync();
    if (!permission.granted) return;
    const result = await ImagePicker.launchImageLibraryAsync({ mediaTypes: ['images'], quality: 0.8 });
    if (result.canceled) return;
    const asset = result.assets[0];
    setBusy(true);
    try {
      const uploaded = await uploadMedia({ uri: asset.uri, name: asset.fileName || 'photo.jpg', type: asset.mimeType || 'image/jpeg' });
      const updated = await updateMe({ [field]: uploaded.url });
      updateUser(updated);
    } finally {
      setBusy(false);
    }
  };

  const save = async () => {
    setSaving(true);
    try {
      const updated = await updateMe({
        name: form.name,
        bio: form.bio,
        work: form.work,
        education: form.education,
        location: form.location,
        links: textToLinks(form.links),
        profileVisibility: form.profileVisibility,
      });
      updateUser(updated);
      navigation.goBack();
    } finally {
      setSaving(false);
    }
  };

  return (
    <Screen>
      <View style={styles.header}>
        <Text style={styles.title}>Edit profile</Text>
        <Pressable onPress={() => navigation.goBack()}>
          <Text style={styles.close}>Cancel</Text>
        </Pressable>
      </View>
      <ScrollView contentContainerStyle={styles.body}>
        <View style={styles.photoRow}>
          <Button title={uploadingAvatar ? 'Uploading...' : 'Change photo'} variant="outline" onPress={() => pickAndUpload('avatarUrl', setUploadingAvatar)} loading={uploadingAvatar} />
          <Button title={uploadingCover ? 'Uploading...' : 'Change cover'} variant="outline" onPress={() => pickAndUpload('coverPhotoUrl', setUploadingCover)} loading={uploadingCover} />
        </View>

        <Input label="Name" placeholder="Your name" value={form.name} onChangeText={set('name')} style={styles.field} />
        <Input label="Bio" placeholder="A short bio" value={form.bio} onChangeText={set('bio')} multiline style={styles.field} />

        <Text style={styles.sectionLabel}>About</Text>
        <Input label="Work" placeholder="e.g. Product Designer at Acme" value={form.work} onChangeText={set('work')} style={styles.field} />
        <Input label="Education" placeholder="e.g. NID Ahmedabad" value={form.education} onChangeText={set('education')} style={styles.field} />
        <Input label="Location" placeholder="e.g. Bengaluru, India" value={form.location} onChangeText={set('location')} style={styles.field} />
        <Input label="Links (one per line)" placeholder="https://example.com" value={form.links} onChangeText={set('links')} multiline style={styles.field} />

        <Text style={styles.sectionLabel}>Who can see your About info</Text>
        <View style={styles.visibilityRow}>
          {VISIBILITY_OPTIONS.map((opt) => (
            <Pressable
              key={opt.id}
              style={[styles.visibilityOption, form.profileVisibility === opt.id && styles.visibilityOptionActive]}
              onPress={() => set('profileVisibility')(opt.id)}
            >
              <Text style={[styles.visibilityOptionText, form.profileVisibility === opt.id && styles.visibilityOptionTextActive]}>{opt.label}</Text>
            </Pressable>
          ))}
        </View>

        <Button title={saving ? 'Saving...' : 'Save'} onPress={save} loading={saving} style={{ marginTop: 20 }} />
      </ScrollView>
    </Screen>
  );
}

const styles = StyleSheet.create({
  header: { flexDirection: 'row', justifyContent: 'space-between', alignItems: 'center', padding: 16, borderBottomWidth: 1, borderBottomColor: colors.line },
  title: { fontSize: 18, fontWeight: '800', color: colors.ink },
  close: { color: colors.accent, fontWeight: '600' },
  body: { padding: 16, paddingBottom: 60 },
  photoRow: { flexDirection: 'row', gap: 10, marginBottom: 16 },
  field: { marginBottom: 14 },
  sectionLabel: { fontSize: 11, fontWeight: '700', color: colors.inkSoft, textTransform: 'uppercase', marginBottom: 10, marginTop: 4 },
  visibilityRow: { flexDirection: 'row', gap: 8 },
  visibilityOption: { flex: 1, paddingVertical: 10, borderRadius: radius.md, borderWidth: 1, borderColor: colors.line, alignItems: 'center' },
  visibilityOptionActive: { backgroundColor: colors.accent, borderColor: colors.accent },
  visibilityOptionText: { fontSize: 13, color: colors.ink, fontWeight: '600' },
  visibilityOptionTextActive: { color: colors.white },
});

File: mobile/lib/features/profile/screens/edit_profile_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:image_picker/image_picker.dart';

import '../../../common/widgets/app_button.dart';
import '../../../common/widgets/app_input.dart';
import '../../../common/widgets/avatar.dart';
import '../../../common/widgets/screen.dart';
import '../../../config/theme.dart';
import '../../../core/auth_session.dart';
import '../../../core/media_api.dart';
import '../../../models/user.dart';
import '../../auth/providers/auth_provider.dart';
import '../data/user_api.dart';

List<UserLink> _textToLinks(String text) {
  return text
      .split('\n')
      .map((line) => line.trim())
      .where((line) => line.isNotEmpty)
      .map((url) {
        var domain = url.replaceFirst(RegExp(r'^https?://'), '');
        domain = domain.split('/').first;
        return UserLink(label: domain, url: url);
      })
      .toList();
}

String _linksToText(List<UserLink> links) => links.map((l) => l.url).join('\n');

/// Ports `src/screens/profile/EditProfileScreen.js`.
class EditProfileScreen extends ConsumerStatefulWidget {
  const EditProfileScreen({super.key});

  @override
  ConsumerState<EditProfileScreen> createState() => _EditProfileScreenState();
}

class _EditProfileScreenState extends ConsumerState<EditProfileScreen> {
  late final TextEditingController _name;
  late final TextEditingController _bio;
  late final TextEditingController _work;
  late final TextEditingController _education;
  late final TextEditingController _location;
  late final TextEditingController _links;
  String _visibility = 'public';
  bool _saving = false;
  bool _uploadingAvatar = false;
  bool _uploadingCover = false;

  @override
  void initState() {
    super.initState();
    final user = ref.read(authSessionProvider).user ?? {};
    _name = TextEditingController(text: user['name'] as String? ?? '');
    _bio = TextEditingController(text: user['bio'] as String? ?? '');
    _work = TextEditingController(text: user['work'] as String? ?? '');
    _education = TextEditingController(text: user['education'] as String? ?? '');
    _location = TextEditingController(text: user['location'] as String? ?? '');
    final links = (user['links'] as List<dynamic>? ?? [])
        .map((e) => UserLink.fromJson(e as Map<String, dynamic>))
        .toList();
    _links = TextEditingController(text: _linksToText(links));
    _visibility = user['profileVisibility'] as String? ?? 'public';
  }

  @override
  void dispose() {
    _name.dispose();
    _bio.dispose();
    _work.dispose();
    _education.dispose();
    _location.dispose();
    _links.dispose();
    super.dispose();
  }

  Future<void> _pickAndUpload({required bool isCover}) async {
    final picker = ImagePicker();
    final picked = await picker.pickImage(source: ImageSource.gallery, imageQuality: 80);
    if (picked == null) return;

    setState(() => isCover ? _uploadingCover = true : _uploadingAvatar = true);
    try {
      final uploaded = await ref.read(mediaApiProvider).upload(
            path: picked.path,
            filename: picked.name,
            mimeType: picked.mimeType,
          );
      final field = isCover ? 'coverPhotoUrl' : 'avatarUrl';
      final updated = await ref.read(userApiProvider).updateMe({field: uploaded['url']});
      await ref.read(authActionsProvider).updateUser(updated);
    } finally {
      if (mounted) setState(() => isCover ? _uploadingCover = false : _uploadingAvatar = false);
    }
  }

  Future<void> _save() async {
    setState(() => _saving = true);
    try {
      final updated = await ref.read(userApiProvider).updateMe({
        'name': _name.text.trim(),
        'bio': _bio.text.trim(),
        'work': _work.text.trim(),
        'education': _education.text.trim(),
        'location': _location.text.trim(),
        'links': _textToLinks(_links.text).map((l) => l.toJson()).toList(),
        'profileVisibility': _visibility,
      });
      await ref.read(authActionsProvider).updateUser(updated);
      if (mounted) Navigator.of(context).pop();
    } finally {
      if (mounted) setState(() => _saving = false);
    }
  }

  @override
  Widget build(BuildContext context) {
    final user = ref.watch(authSessionProvider).user;
    final appUser = user != null ? AppUser.fromJson(user) : null;
    final coverUrl = user?['coverPhotoUrl'] as String?;

    return Screen(
      child: Scaffold(
        backgroundColor: Colors.transparent,
        appBar: AppBar(title: const Text('Edit profile')),
        body: SingleChildScrollView(
          padding: const EdgeInsets.all(20),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: [
              Center(
                child: Stack(
                  alignment: Alignment.bottomRight,
                  children: [
                    Avatar(user: appUser, size: AvatarSize.lg),
                    _PhotoEditButton(loading: _uploadingAvatar, onTap: () => _pickAndUpload(isCover: false)),
                  ],
                ),
              ),
              const SizedBox(height: 8),
              Center(
                child: TextButton(
                  onPressed: _uploadingCover ? null : () => _pickAndUpload(isCover: true),
                  child: Text(
                    coverUrl != null ? 'Change cover photo' : 'Add cover photo',
                    style: const TextStyle(color: AppColors.accent),
                  ),
                ),
              ),
              const SizedBox(height: 16),
              AppInput(label: 'Name', controller: _name),
              const SizedBox(height: 16),
              AppInput(label: 'Bio', controller: _bio, maxLines: 3),
              const SizedBox(height: 16),
              AppInput(label: 'Work', controller: _work),
              const SizedBox(height: 16),
              AppInput(label: 'Education', controller: _education),
              const SizedBox(height: 16),
              AppInput(label: 'Location', controller: _location),
              const SizedBox(height: 16),
              AppInput(label: 'Links (one per line)', controller: _links, maxLines: 4),
              const SizedBox(height: 16),
              const Text('Who can see your About info', style: TextStyle(fontWeight: FontWeight.w600, fontSize: 13)),
              const SizedBox(height: 8),
              Wrap(
                spacing: 8,
                children: [
                  for (final option in const ['public', 'friends', 'only_me'])
                    ChoiceChip(
                      label: Text(option),
                      selected: _visibility == option,
                      onSelected: (_) => setState(() => _visibility = option),
                      selectedColor: AppColors.accentSoft,
                    ),
                ],
              ),
              const SizedBox(height: 24),
              AppButton(label: 'Save', onPressed: _saving ? null : _save, loading: _saving),
            ],
          ),
        ),
      ),
    );
  }
}

class _PhotoEditButton extends StatelessWidget {
  final bool loading;
  final VoidCallback onTap;

  const _PhotoEditButton({required this.loading, required this.onTap});

  @override
  Widget build(BuildContext context) {
    return InkWell(
      onTap: loading ? null : onTap,
      borderRadius: BorderRadius.circular(AppRadius.full),
      child: Container(
        width: 28,
        height: 28,
        decoration: const BoxDecoration(color: AppColors.accent, shape: BoxShape.circle),
        alignment: Alignment.center,
        child: loading
            ? const SizedBox(
                width: 14,
                height: 14,
                child: CircularProgressIndicator(strokeWidth: 2, color: AppColors.white),
              )
            : const Icon(Icons.camera_alt, size: 14, color: AppColors.white),
      ),
    );
  }
}
